src/main/java/org/commcare/cases/instance/IndexedFixtureChildElement.java

```diff
@@ -42,6 +60,171 @@ protected TreeElement cache(QueryContext context) {
+
+        String[] scopeSufficientColumnList = analyseScopeSufficientColumnList(cache);
+
+        if (scopeSufficientColumnList == null) {
+            return null;
+        }
+
+        String[] objectMetadata = getElementMetadata(recordId, scopeSufficientColumnList, context);
+
+        partialMatch = this.buildPartialElementFromMetadata(scopeSufficientColumnList, objectMetadata);
+        cache.cacheElement(this.getInstanceName(), recordId, partialMatch);
+        return partialMatch;
+    }
+
+    protected String[] getElementMetadata(int recordId, String[] metaFields, QueryContext context) {
+        if (context == null || this.parent.getStorageCacheName() == null) {
+            return parent.storage.getMetaDataForRecord(recordId, metaFields);
+        }
+        RecordSetResultCache recordSetCache = context.getQueryCacheOrNull(RecordSetResultCache.class);
+
+        String recordSetKey = parent.getStorageCacheName();
+
+        String recordObjectKey = parent.getStorageCacheName() + "_partial";
+
+        RecordObjectCache<String[]> recordObjectCache = StorageInstanceTreeElement.getRecordObjectCacheIfRelevant(
+                context);
+
+        if (recordObjectCache != null) {
+            if (recordObjectCache.isLoaded(recordObjectKey, recordId)) {
+                return recordObjectCache.getLoadedRecordObject(recordObjectKey, recordId);
+            }
+
+            if (StorageInstanceTreeElement.canLoadRecordFromGroup(recordSetCache, recordSetKey, recordId)) {
+                Pair<String, LinkedHashSet<Integer>> tranche =
+                        recordSetCache.getRecordSetForRecordId(recordSetKey, recordId);
+
+                EvaluationTrace loadTrace =
+                        new EvaluationTrace(String.format("Model [%s]: Limited Scope Partial Bulk Load [%s}",
+                                recordObjectKey,tranche.first));
+
+                LinkedHashSet<Integer>  body = tranche.second;
+                parent.getStorage().bulkReadMetadata(body, metaFields, recordObjectCache.getLoadedCaseMap(recordObjectKey));
+                loadTrace.setOutcome("Loaded: " + body.size());
+
+                context.reportTrace(loadTrace);
+
+                return recordObjectCache.getLoadedRecordObject(recordObjectKey, recordId);
+            }
+        }
+
+        return parent.storage.getMetaDataForRecord(recordId, metaFields);
+    }
+
+    private String[] analyseScopeSufficientColumnList(ScopeLimitedReferenceRequestCache cache) {
+        String[] limitedScope = cache.getInternalScopedLimit(this.getInstanceName());
+        if (limitedScope != null) {
+            return limitedScope;
+        }
+
+        //If we don't already have that list, build it (or detect that it's won't be possible and tell
+        //the cache to not try.
+
+        Set<TreeReference> referencesInScope =
+                cache.getInScopeReferences(this.getInstanceName());
+
+        StorageIndexedTreeElementModel model = parent.getModelTemplate();
```

Comment:
It would be nice if the code from here through line 191 was moved into a helper method named something like `checkIfRequiredReferencesAreAllIndexed()` or something to that effect. I'm envisioning that the method would take in the `referencesInScope` argument above, and either return the populated `columnNameCacheLoads` object if it was successful, or `null` if it failed.

Reply:
refactored this below to be more explict about what's actually taking care of what things